sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -212,13 +212,12 @@ private static <T> Mono<Pair<List<Range<String>>, QueryInfo>> getTargetRangesFro
                 queryInfo));
     }
 
-    private static void tryCacheQueryPlan(
+    synchronized private static void tryCacheQueryPlan(
         SqlQuerySpec query,
         PartitionedQueryExecutionInfo partitionedQueryExecutionInfo,
         Map<String, PartitionedQueryExecutionInfo> queryPlanCache) {
-        QueryInfo queryInfo = partitionedQueryExecutionInfo.getQueryInfo();
-        if (canCacheQuery(queryInfo) && !queryPlanCache.containsKey(query.getQueryText())) {
-            if (queryPlanCache.size() == Constants.QUERYPLAN_CACHE_SIZE) {
+        if (canCacheQuery(partitionedQueryExecutionInfo.getQueryInfo()) && !queryPlanCache.containsKey(query.getQueryText())) {
+            if (queryPlanCache.size() >= Constants.QUERYPLAN_CACHE_SIZE) {
```

Comment:
Ah - now i see. IMO this is the only change necessary. Much less noise, less risk to perf changes and also less confusing to use different synchronization mechanisms (ConcurrentHashMap but still using synchronized)

The comparison above is simply a bug using >= instead will fix the issue sufficiently. IMO this should be the only change in this PR

Reply:
Actually, we thought of going with that change initially, but based on further discussion and brainstorming, we didn't. 

If we don't synchronize the whole method, different threads will still read different values for cache size, and all of them will try to clear the cache. This can cause cache to remove entries added by other threads during this time. 
And based on the test that I have added, when we repro this locally, multiple threads read same values and multiple threads clear the cache unnecessarily removing existing values from cache which the other threads are adding. 

Also, I have done perf tests and there is no perf issue at all whatsoever based on the results. So that's why IMO, we can go with the synchronized change.

Reply:
Above race conditions aren't critical - but it brings up a good point because it isn't obvious that those races aren't critical. And given that there is no I/O happening in this now synchronized method it seems like a good fix. Thanks for the explanation!